src/main/java/de/tum/in/www1/artemis/service/connectors/localvc/LocalVCService.java

```diff
@@ -0,0 +1,281 @@
+        return new LocalVCRepositoryUrl(localVCServerUrl, projectKey, courseShortName, repositorySlug);
+    }
+
+    @Override
+    public void setRepositoryPermissionsToReadOnly(VcsRepositoryUrl repositoryUrl, String projectKey, Set<User> users) throws LocalVCException {
+        // Not implemented for local VC. All checks for whether a student can access a repository are conducted in the LocalVCFetchFilter and LocalVCPushFilter.
+    }
+
+    /**
+     * Get the default branch of the repository
+     *
+     * @param repositoryUrl The repository url to get the default branch for.
+     * @return the name of the default branch, e.g. 'main'
+     */
+    @Override
+    public String getDefaultBranchOfRepository(VcsRepositoryUrl repositoryUrl) throws LocalVCException {
+        try {
+            LocalVCRepositoryUrl localVCRepositoryUrl = new LocalVCRepositoryUrl(localVCServerUrl, repositoryUrl.toString());
+            String localRepositoryPath = localVCRepositoryUrl.getLocalPath(localVCPath).toString();
+            Map<String, Ref> remoteRepositoryRefs = Git.lsRemoteRepository().setRemote(localRepositoryPath).callAsMap();
+            if (remoteRepositoryRefs.containsKey("HEAD")) {
+                return remoteRepositoryRefs.get("HEAD").getTarget().getName();
+            }
+
+            throw new LocalVCException("Cannot get default branch of repository " + localRepositoryPath + ". ls-remote does not return a HEAD reference.");
+        }
+        catch (Exception e) {
+            log.error("Unable to get default branch for repository {}", repositoryUrl.folderNameForRepositoryUrl(), e);
+            throw new LocalVCException("Unable to get default branch for repository " + repositoryUrl.folderNameForRepositoryUrl(), e);
+        }
+    }
+
+    @Override
+    public void unprotectBranch(VcsRepositoryUrl repositoryUrl, String branch) throws VersionControlException {
+        // Not implemented. It is not needed for local VC for the current use
+        // case, because the main branch is unprotected by default.
+    }
+
+    /**
+     * Check if a project already exists in the file system to make sure the new projectKey is unique.
+     *
+     * @param projectKey  to check if a project with this unique key already exists.
+     * @param projectName to check if a project with the same name already exists.
+     * @return true or false depending on whether the respective folder exists.
+     */
+    @Override
+    public boolean checkIfProjectExists(String projectKey, String courseShortName, String projectName) {
+        String projectKeyStripped = StringUtil.stripIllegalCharacters(projectKey);
+        String courseShortNameStripped = StringUtil.stripIllegalCharacters(courseShortName);
+
+        // Try to find the folder in the file system. If it is not found, return false.
+        Path projectPath = Path.of(localVCPath, courseShortNameStripped, projectKeyStripped);
```

Comment:
Doesn't the project key already contain some version of the course short name?

Reply:
That is true. I added the course short name because it made authorization with the old authorization implementation easier and because this way I had a dedicated folder for each course which holds one folder for each exercise instead of having one folder `local-vcs-repos` that holds one folder for each exercise not split up by courses. I guess that does not justify the numerous changes to the code though so I changed it to only use the projectKey as the identifier.